### PR TITLE
Configurable providers

### PR DESCRIPTION
**Pull request by woksin**

### CODE REVIEW

File: Source/Login/.docker/config.yaml [hunk 1]
```diff
@@ -4,6 +4,14 @@ serve:
...
   error: http://localhost:8080/error
 
+providers:
+  microsoft:
+    name: Microsoft
+    image_url: https://aadcdn.msauth.net/shared/1.0/content/images/microsoft_logo_ed9c9eb0dce17d752bedea6b5acda6d9.png
+  github:
+    name: GitHub
+    image_url: https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png
+
```
Comment (jakhog) [inline Source/Login/.docker/config.yaml]: I think this should be empty by default
```suggestion
providers: {}

```

File: Source/Login/Web/select-provider/IdentityProviderListItem.tsx [hunk 2]
```diff
@@ -25,7 +25,8 @@ export class IdentityProviderListItem extends React.Component<IdentityProviderLi
             <DocumentCard type={DocumentCardType.compact} onClick={this.onClick}>
                 <DocumentCardPreview previewImages={[
                     {
-                        previewImageSrc: 'https://docs.microsoft.com/en-us/azure/active-directory/develop/media/howto-add-branding-in-azure-ad-apps/ms-symbollockup_mssymbol_19.svg',
+                        // previewImageSrc: 'https://docs.microsoft.com/en-us/azure/active-directory/develop/media/howto-add-branding-in-azure-ad-apps/ms-symbollockup_mssymbol_19.svg',
+                        previewImageSrc: this.props.provider.imageURL || 'https://docs.microsoft.com/en-us/azure/active-directory/develop/media/howto-add-branding-in-azure-ad-apps/ms-symbollockup_mssymbol_19.svg',
```
Comment (jakhog) [inline Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Maybe we just don't show an image if the imageURL is not set?

File: Source/Login/Backend/providers/getter.go [hunk 3]
```diff
@@ -4,15 +4,34 @@ type Getter interface {
...
+type getter struct {
+	configuration Configuration
+}
 
 func (g *getter) GetProviderByID(id string) (Provider, error) {
+	provider, ok := g.configuration.Providers()[id]
+	if !ok {
+		return Provider{
+			ID:       id,
+			Display:  id + " Display Name",
```
Comment (jakhog) [inline Source/Login/Backend/providers/getter.go]: This was just for testing purposes, we can just get rid of it - it looks bad...
```suggestion
			Display:  id,
```

File: Source/Login/Backend/configuration/viper/providers_configuration.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	imageURL, err := url.Parse(imageURLString)
+	if err != nil {
+		return newNilProvider()
+	}
+
+	return &providers.ProviderConfiguration{
+		Name:     name,
+		ImageURL: imageURL,
+	}
```
Comment (jakhog) [inline Source/Login/Backend/configuration/viper/providers_configuration.go]: In stead of returning the default NilProvider thing, I suggest also returning an error or just a bool from this function in addition to the provider, and if it is not all ok => don't add it to the map.

File: Source/Login/Backend/configuration/viper/providers_configuration.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+func (c *providersConfiguration) Providers() providers.Providers {
+	if !viper.IsSet(providersKey) {
+		return nil
+	}
+	return getProviderConfigurationMap(viper.GetStringMap(providersKey))
+}
+
+func getProviderConfigurationMap(in map[string]interface{}) providers.Providers {
+	providerConfigMap := make(map[providers.ProviderID]*providers.ProviderConfiguration, len(in))
```
Comment (jakhog) [inline Source/Login/Backend/configuration/viper/providers_configuration.go]: I don't think we need to be fancy to save memory allocations with the size:
```suggestion
	providerConfigMap := make(map[providers.ProviderID]*providers.ProviderConfiguration)
```

**Review by jakhog — CHANGES_REQUESTED** [review]

File: Source/Login/Web/select-provider/IdentityProviderListItem.tsx (hunk 2 above)
Comment (woksin) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: So empty string? Or not set the property?

File: Source/Login/Backend/configuration/viper/providers_configuration.go (hunk 5 above)
Comment (woksin) [reply Source/Login/Backend/configuration/viper/providers_configuration.go]: 👍

Comment (woksin) [reply Source/Login/Backend/configuration/viper/providers_configuration.go]: Fixed by https://github.com/dolittle-platform/Authentication/pull/10/commits/68d3ca5afdfe219c74055ab494857bcdec2339be

File: Source/Login/Backend/configuration/viper/providers_configuration.go (hunk 4 above)
Comment (woksin) [reply Source/Login/Backend/configuration/viper/providers_configuration.go]: Fixed by https://github.com/dolittle-platform/Authentication/pull/10/commits/68d3ca5afdfe219c74055ab494857bcdec2339be

File: Source/Login/Web/select-provider/IdentityProviderListItem.tsx (hunk 2 above)
Comment (woksin) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Will this fix it? https://github.com/dolittle-platform/Authentication/pull/10/commits/6c603490e50540c7fda28d626abfb64db4152004

Comment (woksin) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: It worked when I tested it locally

Comment (jakhog) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Tested it just now myself, it complained about an unknown source. I have another suggestion

Comment (woksin) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Weird. Then come with it :P

File: Source/Login/Web/select-provider/IdentityProviderListItem.tsx [hunk 6]
```diff
@@ -25,7 +25,8 @@ export class IdentityProviderListItem extends React.Component<IdentityProviderLi
             <DocumentCard type={DocumentCardType.compact} onClick={this.onClick}>
                 <DocumentCardPreview previewImages={[
```
Comment (jakhog) [inline Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Ahh, I can't make suggestions, but here is what we could do to not show the preview if no image is set:
```typescript
                { this.props.provider.imageURL &&
                    <DocumentCardPreview previewImages={[
                        {
                            previewImageSrc: this.props.provider.imageURL,
                            width: 60,
                            height: 60,
                        }
                    ]} />
                }
```

Comment (woksin) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Is this the code you're looking for https://github.com/dolittle-platform/Authentication/pull/10/commits/a4ee3159909061b366706f7840bc46737ad3ec2f ?

Comment (jakhog) [reply Source/Login/Web/select-provider/IdentityProviderListItem.tsx]: Ohh, yes!

**Review by jakhog — APPROVED** [review]

**Pull request merged**